File: Paradigmas_de_Programacao/Atividade_9/Ex8/Matriz.hs
module Matriz (Row, Matriz, Length, sumElementMatriz, sumTwoMatriz, verifyNumbers) where

type Row = [Int]
type Matriz = [Row]
type Length = Int

sumRow :: Row -> Int
sumRow [] = 0
sumRow (a:b) = a + sumRow b

sumElementMatriz :: Matriz -> Int
sumElementMatriz [] = 0
sumElementMatriz (a:b) = (sumRow a) + (sumElementMatriz b)

sumTwoMatriz :: Matriz -> Matriz -> Matriz
sumTwoMatriz a b = zipWith (zipWith (+)) a b

verifyNumbers :: Matriz -> Bool
verifyNumbers m = verifyNumbersAux (flatten m) (lenRow (head m)) 1

verifyNumbersAux :: Row -> Length -> Int -> Bool
verifyNumbersAux m l x
    | (x > l*l) = True
    | ((x <= l*l) && (hasNumber x m)) = verifyNumbersAux m l (x+1)
    | otherwise = False


lenRow :: Row -> Int
lenRow [] = 0
lenRow (a:b) = 1 + lenRow b

hasNumber :: Int -> Row -> Bool
hasNumber x [] = False
hasNumber x (a:b) 
    | (x == a) = True
    | otherwise = hasNumber x b

flatten :: Matriz -> Row
flatten a = (\z n -> foldr (\x y -> foldr z y x) n a) (:) []

File: INE5202/Teste/matriz.m
function [A, B] = matriz(n)

	n1 = 5;
	n2 = 10;
	n3 = 20;

	A = zeros(n, n);
	B = zeros(n, 1);


	i = 1;
	A(i,i) = 3;
	A(i,(i+2)) = 2;
	B(i,1) = 2;

	for (i = 2:n1)
		A(i,(i-1)) = -2;
		A(i,i) = 4;
		A(i,(i+5)) = -3;
		B(i,1) = -3;
	end

	for (i = (n1+1):n2)
		A(i,(i-4)) = -8;
		A(i,i) = -4;
		A(i,(i+1)) = -3;
		B(i,1) = 3;
	end

	for (i = (n2+1):n3)
		A(i,(i-3)) = -3;	
		A(i,(i-1)) = -2;
		A(i,i) = 7;
		B(i,1) = 5;
	end


end
